cassandra/cluster.py

```diff
@@ -3722,10 +3722,14 @@ def _on_timeout(self, _attempts=0):
         if self._connection is not None:
             try:
                 self._connection._requests.pop(self._req_id)
-            # This prevents the race condition of the
-            # event loop thread just receiving the waited message
-            # If it arrives after this, it will be ignored
+            # PYTHON-1044
+            # This request might have been removed from the connection after the latter was defunct by heartbeat.
+            # We should still raise OperationTimedOut to reject the future so that the main event thread will not
+            # wait for it endlessly
             except KeyError:
+                key = "Connection defunct by heartbeat"
+                errors = {key: "Client request timeout. See Session.execute[_async](timeout)"}
+                self._set_final_exception(OperationTimedOut(errors, self._current_host))
```

Comment:
Comment out to dead lock with unit and integration tests

Reply:
That's what I meant by `without your fix`. Trying to reproduce the deadlock with simulacron...

Reply:
So I guess it's not deadlocking for you ? I tried on several machines with a freshly cloned repository using `simulacron-standalone-0.8.9.jar`. I already noticed that it's deadlocking only 99% of the time. Can you tell me more about your set-up ?

Is your simularcron configuration creating a cluster of 3 hosts (127.0.0.1-3) ? Maybe you can try to increase `lag = 0.05` to `lag = 0.1` or `lag = 0.2` within the integration test.

Reply:
Thanks... I will make sure I'm using the same simulacron version and vary the lag to see.

Reply:
What event loop are you using? asyncore, libev? What Python 2.7 version? Could you also attach the stacktrace of this test when it deadlocks?

Reply:
Looks like I can reproduce the deadlock in our jenkins CI platform, but not locally. Better than nothing...